Downport: 1.20.1

MianReplicate opened the pull request:
This PR aims to port the current master branch 1.20.2 version of TR to 1.20.1 in full working state for both Neo (hopefully) and Fabric.

Can I just say that I am suffering? This port forced me to look into what a codec was to fix one (and I am 99% sure it's not fixed!)

MianReplicate committed:
forgot to icnlude this in last coimmit

MianReplicate committed:
fabric now have resources again from forge yay

Jeryn99 committed:
ImageButton my beloved

Jeryn99 committed:
The UI torture is over

Jeryn99 committed:
Fixed the ManipulatorCraftingRecipeSerializer

Jeryn99 committed:
Unlock all and lock all commands

Olivier1152005 commented:
Is there an easy way to download the decompiled version of the mod that is on GitHub without having to manually go to each file, download it, and then manually rebuild every file and path? Is there anything in particular you are having trouble with? I like this mod and want to help get it ported to 1.20.1.

CommandrMoose commented:
> Is there an easy way to download the decompiled version of the mod that is on GitHub without having to manually go to each file, download it, and then manually rebuild every file and path? Is there anything in particular you are having trouble with? I like this mod and want to help get it ported to 1.20.1.

I recommend looking into how git works. This PR is close to coming to a close.
https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository

Jeryn99 committed:
Fixed Submit button using a ImageButton that doesn't actually use an image

Jeryn99 commented:
> Is there an easy way to download the decompiled version of the mod that is on GitHub without having to manually go to each file, download it, and then manually rebuild every file and path? Is there anything in particular you are having trouble with? I like this mod and want to help get it ported to 1.20.1.

Hello. What you are viewing is code that is cutting edge. What I mean by that is that it is so newly published that if you were to cut the review and dev process out and ran off with these files, you would end up destroying your Minecraft saves. This is a pull request for developers for tracking the port downwards and if you do not know what you are doing, you should not interact as you are hindering the process. Think of it as walking into a car factory and asking to take the car in its current state (Engine, no brakes yet though) because you love it, just wouldn't go down well

MianReplicate committed:
i wanted to feel like i contributed so i ported ported the fixes from WhoCraft/TardisRefined#267

50ap5ud5 commented:
PR source branch needs to be switched to a recently created branch called [minecraft/1.20.1_port](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1_port) before being moved out of a draft state.

It is a copy of the porting work but rebased on top of the latest 20.2 changes. This means this PR would up to date with the latest 1.20.2 changes. Those reviewing the PR would only be seeing changes that were made during the port.

Jeryn99 commented:
This does sound like we need to make a whole new PR?

50ap5ud5 commented:
> This does sound like we need to make a whole new PR?

Actually, I think you are right here. 

I had made a new branch in case I significantly changed the commit history. It looks it wasn't the case, so if we want to continue using this PR, we just need to override the [LoxiGoose:minecraft/1.20.1](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1) branch with that of the [LoxiGoose:minecraft/1.20.1_port](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1_port) branch.

50ap5ud5 commented:
Ok, I've updated [LoxiGoose:minecraft/1.20.1](https://github.com/LoxiGoose/TardisRefined-Ports/tree/minecraft/1.20.1) to be rebased on top of [WhoCraft:minecraft/1.20](https://github.com/WhoCraft/TardisRefined/tree/minecraft/1.20) changes so that we don't need to make a new PR.

Jeryn99 commented:
we're really damn cooking

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java:
```diff
@@ -15,12 +17,7 @@ public abstract class DeferredRegistry<T> {
     public abstract <R extends T> RegistrySupplier<R> register(String id, Supplier<R> supplier);
 
-    /** Create a RegistrySupplierHolder. This has similar behaviour to a Supplier but has vanilla Holder attributes such as use in tags*/
-    public abstract <I extends T> RegistrySupplierHolder<T, I> registerHolder(final String name, final Supplier<I> sup);
```
Why was the ability to register via a Holder removed? The reason it was added in 20.2 was because a ``Holder`` is a vanilla equivalent of our ``RegistrySupplier``. I had based our version off Neoforge's DeferredHolder to allow for us to start to move towards using more vanilla systems so that it will be easier to detect if issues came from us or vanilla's upstream logic.

50ap5ud5 replied:
Forge's registry system in 1.20.1 still retains an older design which was completely isolated from vanilla. The inability to access the underlying vanilla registry and vanilla objects such as Holders make this unfeasible unfortunately.

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java:
```diff
@@ -15,12 +17,7 @@ public abstract class DeferredRegistry<T> {
-
-    /** Get the underlying registry, which includes all entries added by any mod that has a DeferredRegistry with the same ResourceKey.
-     * All lookup methods should be called from here.*/
-    public abstract Registry<T> getRegistry();
```
Why was the ability to get the underlying Registry removed? This was added in 20.2 so that we can utilise methods specific to the vanilla `Registry<T>` object.

50ap5ud5 replied:
Forge's registry system in 1.20.1 still retains an older design which was completely isolated from vanilla. The inability to access the underlying vanilla registry make this method unfeasible unfortunately. Instead, we need to expose lookup methods such as keySet and entrySet manually, so that a common interface is present for both Fabric and Forge experiences.

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/registry/DeferredRegistry.java:
```diff
@@ -34,19 +31,7 @@ public static <T> DeferredRegistry<T> create(String modid, ResourceKey<? extends
     }
 
-    /** Gets the underlying Codec for the registry object type, if defined. Currently unused and untested, but added for completeness*/
-    public abstract Supplier<Codec<T>> getCodec();
```
Why was this limited to only the `CustomRegistry`? `getCodec` was added in 20.2 for all of our mod's registries because many registries in vanilla use an underlying codec, such as BlockStates and ItemStacks.

If in the future we needed to create any data-driven features this method would have allowed us to retrieve the codec to use for data read/writing.

CommandrMoose commented on common/src/main/java/whocraft/tardis_refined/client/screen/selections/ShellSelectionScreen.java:
```diff
@@ -152,10 +154,10 @@ public void render(GuiGraphics guiGraphics, int i, int j, float f) {
-    public void renderBackground(GuiGraphics guiGraphics, int i, int j, float f) {
-        // super.renderBackground(guiGraphics, i, j, f);
-    }
+//    @Override
```
Remove comments

50ap5ud5 commented on forge/src/main/java/whocraft/tardis_refined/registry/forge/CustomRegistryImpl.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public static <T> CustomRegistry<T> create(ResourceLocation id, boolean syncToClient) {
+        DeferredRegister<T> deferredRegister = DeferredRegister.create(id, id.getNamespace());
```
This implementation looks to be quite different from upstream 20.2. Any reason for this refactor?

Here `CustomRegistry` is creating a new Forge `DeferredRegister` instance, and later in `DeferredRegistryImpl`, we are creating a duplicate Forge `DeferredRegister` by calling `DeferredRegister#create` a second time.

Ideally we should have one `DeferredRegistry` instance, so that it simplifies the code for addon mod developers, any lookup operations (`getKeys`, `getValues`) would be looking from a single source of truth.
Additionally, it would reduce developer confusion when deciding whether to use `CustomRegistry` or `DeferredRegistry` for lookup values, and reduce maintenance efforts between 20.1 and 20.2.

50ap5ud5 commented on fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java:
```diff
@@ -22,35 +24,19 @@ public static <T> DeferredRegistry<T> create(String modid, ResourceKey<? extends
-        private final boolean isCustom;
-        private final boolean syncToClient;
         private ResourceKey<? extends Registry<T>> registryKey;
+        private final List<RegistrySupplier<T>> entries;
```
`getEntries` should not be looking up from a separate list of entries. We need to be using the source of truth, which is the underlying `WritableRegistry<T>` for any lookups, because only that registry will contain any new additions added by addon mods.

50ap5ud5 replied:
Fixed by downporting 1.20.1 registry systems which use the underlying registry as the source of truth

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/registry/RegistrySupplierHolder.java:
```diff
@@ -1,228 +0,0 @@
- * <br> Based off NeoForge's DeferredHolder
- * <br> https://github.com/neoforged/NeoForge/blob/1.20.x/src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java
- * */
-public class RegistrySupplierHolder<R, T extends R> implements Holder<R>, Supplier<T> {
```
Why was this removed? It was kept as an alternative way to register objects which need to use vanilla features such as the ability to specify in datapack tags.

50ap5ud5 commented on fabric/src/main/java/whocraft/tardis_refined/registry/fabric/DeferredRegistryImpl.java:
```diff
@@ -62,25 +48,13 @@ public void register() {
-        public Supplier<Codec<T>> getCodec() {
-            return () -> this.registry.byNameCodec();
+        public List<RegistrySupplier<T>> getEntries() {
+            return this.entries;
```
Implementing these methods is good for convenience purposes, but it needs to call the relevant methods from the underlying `Registry<T>`. It's needed to ensure that we are looking at the list of entries which addon mods have the ability to add to.

50ap5ud5 commented on forge/src/main/java/whocraft/tardis_refined/registry/forge/DeferredRegistryImpl.java:
```diff
@@ -0,0 +1,62 @@
+        private Registry<T> registry;
+
+        private ResourceKey<? extends Registry<T>> registryKey;
+        private final List<RegistrySupplier<T>> entries;
```
`getEntries` should not be looking up from a separate list of entries. We need to be using the source of truth, which is the underlying `WritableRegistry<T>` for any lookups, because only that registry will contain any new additions added by addon mods.

50ap5ud5 commented on forge/src/main/java/whocraft/tardis_refined/registry/forge/DeferredRegistryImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+        @Override
+        public List<RegistrySupplier<T>> getEntries() {
+            return this.entries;
```
Implementing these methods is good for convenience purposes, but it needs to call the relevant methods from the underlying `Registry<T>`. It's needed to ensure that we are looking at the list of entries which addon mods have the ability to add to.

CommandrMoose commented on common/src/main/java/whocraft/tardis_refined/client/screen/upgrades/UpgradeWidget.java:
```diff
@@ -194,68 +196,59 @@ public void drawHover(GuiGraphics guiGraphics, int i, int j, float f, int k, int
-        guiGraphics.blitSprite(UpgradesScreen.getFrame(upgradeEntry.getUpgradeType(), upgradeEntry.isUnlocked(upgradeHandler)), i + this.x + 3, j + this.y, 26, 26);
+        //blit(ResourceLocation atlasLocation, int x, int y, float uOffset, float vOffset, int width, int height, int textureWidth, int textureHeight)
+
+      //  guiGraphics.blit(UpgradesScreen.getBox(upgradeEntry.isUnlocked(upgradeHandler)), i + this.x + 3, j + this.y, 0,0, 200, 26,200, 26);
```
Remove comments

CommandrMoose commented on common/src/main/java/whocraft/tardis_refined/client/screen/waypoints/WaypointListScreen.java:
```diff
@@ -153,18 +154,18 @@ public GenericMonitorSelectionList createSelectionList() {
-        // super.renderBackground(guiGraphics, i, j, f);
-    }
+//    @Override
+//    public void renderBackground(GuiGraphics guiGraphics, int i, int j, float f) {
```
Going to write it for the final time here. There's lots around the place

CommandrMoose reviewed (changes requested):
Comments either need to be removed or please tell my why they should stay.

There's a few changes on 1.20.2's branch that should be cherry picked in, but then I'd say it is GTG

50ap5ud5 commented on common/src/main/java/whocraft/tardis_refined/registry/CustomRegistry.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Set;
+import java.util.function.Supplier;
+
+public abstract class CustomRegistry<T> {
```
This implementation looks to be quite different from upstream 20.2. Any reason for this refactor?

50ap5ud5 reviewed (changes requested):
This is amazing work. The refactors made in the registry classes need changes as they currently prevent the ability for add-on mods to add new entries to TARDIS Refined. It also appears to remove previous functionality, some of which can be addressed with some better documentation.

I have incorporated some of your changes into the upstream 20.2 branch so that it is closer to this implementation, but generally it would be easier for maintenance if we kept similar implementation across versions where possible.

Jeryn99 commented:
> This is amazing work. The refactors made in the registry classes need changes as they currently prevent the ability for add-on mods to add new entries to TARDIS Refined. It also appears to remove previous functionality, some of which can be addressed with some better documentation.

When getting 1.20.1 running, we ripped out all the registry stuff in favour of CustomRegistry from Regeneration/Weeping Angels as a temporary measure as the intention was to get the mod running and see what we would need to fix at runtime. CustomRegistry was always a temporary measure and I should have documented it as so. The intention post fixes was to have your registry refactor ported down

Jeryn99 committed:
Removed commented code

Jeryn99 commented:
> Comments either need to be removed or please tell my why they should stay.
> 
> There's a few changes on 1.20.2's branch that should be cherry picked in, but then I'd say it is GTG

All commented code has now been removed from the mod where not needed

![image](https://github.com/WhoCraft/TardisRefined/assets/36424503/a9a11841-1a21-4e30-a31d-6c020820e836)

MianReplicate committed:
Progress I guess?

MianReplicate committed:
forge mappings are fixed omg omg yes i swapped to forge, no fuck you im not swapping back

Jeryn99 committed:
Can you imagine how much money you would lose on this?

Jeryn99 committed:
It compiles

MianReplicate committed:
Finally.. after millions of registry shit, we can now boo- and now the code is complaining about patterns not existing in the registries. FUCK.

Jeryn99 committed:
bam boom everything works but UI

Jeryn99 committed:
Work on UIS, not done, SpiteButtons need ported correctly

50ap5ud5 committed:
Remove CustomRegistry in favour of keeping current implementation

50ap5ud5 committed:
Registry downport

-Regression: Disabled the getRegistry in order to maintain a common modding interface due to Forge registries being isolated from vanilla, and not exposing the vanilla registry
-Regression: Disabled the ability to use a vanilla Holder for registration due to Forge registries being isolated from vanilla, and not exposing the vanilla registry
-Moved default values for Patterns and data-driven registries to a method so that they can be called at a specific game loading stage due to specific modloader behaviours.

CommandrMoose committed:
Watchface desktop and Aviatrax ambiance
